File: frontend/src/api/admin/tlsFingerprintProfile.ts
/**
 * Admin TLS Fingerprint Profile API endpoints
 * Handles TLS fingerprint profile CRUD for administrators
 */

import { apiClient } from '../client'

export interface HTTP2ProfileConfig {
  initial_window_size?: number
  connection_window_update?: number
  max_header_list_size?: number
  enable_push?: boolean
}

export interface TLSProfileDefaults {
  openai_oauth_default_tls_profile_id: number
}

/**
 * TLS fingerprint profile interface
 */
export interface TLSFingerprintProfile {
  id: number
  name: string
  description: string | null
  enable_grease: boolean
  shuffle_extensions: boolean
  http2: HTTP2ProfileConfig | null
  cipher_suites: number[]
  curves: number[]
  point_formats: number[]
  signature_algorithms: number[]
  alpn_protocols: string[] | null
  supported_versions: number[]
  key_share_groups: number[]
  psk_modes: number[]
  extensions: number[]
  created_at: string
  updated_at: string
}

/**
 * Create profile request
 */
export interface CreateProfileRequest {
  name: string
  description?: string | null
  enable_grease?: boolean
  shuffle_extensions?: boolean
  http2?: HTTP2ProfileConfig | null
  cipher_suites?: number[]
  curves?: number[]
  point_formats?: number[]
  signature_algorithms?: number[]
  alpn_protocols?: string[] | null
  supported_versions?: number[]
  key_share_groups?: number[]
  psk_modes?: number[]
  extensions?: number[]
}

/**
 * Update profile request
 */
export interface UpdateProfileRequest {
  name?: string
  description?: string | null
  enable_grease?: boolean
  shuffle_extensions?: boolean
  http2?: HTTP2ProfileConfig | null
  cipher_suites?: number[]
  curves?: number[]
  point_formats?: number[]
  signature_algorithms?: number[]
  alpn_protocols?: string[] | null
  supported_versions?: number[]
  key_share_groups?: number[]
  psk_modes?: number[]
  extensions?: number[]
}

export async function list(): Promise<TLSFingerprintProfile[]> {
  const { data } = await apiClient.get<TLSFingerprintProfile[]>('/admin/tls-fingerprint-profiles')
  return data
}

export async function getById(id: number): Promise<TLSFingerprintProfile> {
  const { data } = await apiClient.get<TLSFingerprintProfile>(`/admin/tls-fingerprint-profiles/${id}`)
  return data
}

export async function create(profileData: CreateProfileRequest): Promise<TLSFingerprintProfile> {
  const { data } = await apiClient.post<TLSFingerprintProfile>('/admin/tls-fingerprint-profiles', profileData)
  return data
}

export async function update(id: number, updates: UpdateProfileRequest): Promise<TLSFingerprintProfile> {
  const { data } = await apiClient.put<TLSFingerprintProfile>(`/admin/tls-fingerprint-profiles/${id}`, updates)
  return data
}

export async function deleteProfile(id: number): Promise<{ message: string }> {
  const { data } = await apiClient.delete<{ message: string }>(`/admin/tls-fingerprint-profiles/${id}`)
  return data
}

export async function parseYAML(yaml: string): Promise<CreateProfileRequest> {
  const { data } = await apiClient.post<CreateProfileRequest>('/admin/tls-fingerprint-profiles/parse-yaml', { yaml })
  return data
}

export async function getDefaults(): Promise<TLSProfileDefaults> {
  const { data } = await apiClient.get<TLSProfileDefaults>('/admin/tls-fingerprint-profiles/defaults')
  return data
}

export async function setDefaults(defaults: TLSProfileDefaults): Promise<TLSProfileDefaults> {
  const { data } = await apiClient.put<TLSProfileDefaults>('/admin/tls-fingerprint-profiles/defaults', defaults)
  return data
}

export const tlsFingerprintProfileAPI = {
  list,
  getById,
  create,
  update,
  delete: deleteProfile,
  parseYAML,
  getDefaults,
  setDefaults
}

export default tlsFingerprintProfileAPI

File: backend/internal/model/tls_fingerprint_profile.go
// Package model 定义服务层使用的数据模型。
package model

import (
	"time"

	"github.com/Wei-Shaw/sub2api/internal/pkg/tlsfingerprint"
)

// TLSFingerprintProfile TLS 指纹配置模板
// 包含完整的 ClientHello 参数，用于模拟特定客户端的 TLS 握手特征
type TLSFingerprintProfile struct {
	ID                  int64                       `json:"id"`
	Name                string                      `json:"name"`
	Description         *string                     `json:"description"`
	EnableGREASE        bool                        `json:"enable_grease"`
	ShuffleExtensions   bool                        `json:"shuffle_extensions"`
	HTTP2               *tlsfingerprint.HTTP2Config `json:"http2"`
	CipherSuites        []uint16                    `json:"cipher_suites"`
	Curves              []uint16                    `json:"curves"`
	PointFormats        []uint16                    `json:"point_formats"`
	SignatureAlgorithms []uint16                    `json:"signature_algorithms"`
	ALPNProtocols       []string                    `json:"alpn_protocols"`
	SupportedVersions   []uint16                    `json:"supported_versions"`
	KeyShareGroups      []uint16                    `json:"key_share_groups"`
	PSKModes            []uint16                    `json:"psk_modes"`
	Extensions          []uint16                    `json:"extensions"`
	CreatedAt           time.Time                   `json:"created_at"`
	UpdatedAt           time.Time                   `json:"updated_at"`
}

// Validate 验证模板配置的有效性
func (p *TLSFingerprintProfile) Validate() error {
	if p.Name == "" {
		return &ValidationError{Field: "name", Message: "name is required"}
	}
	if err := p.ToTLSProfile().Validate(); err != nil {
		return &ValidationError{Field: "profile", Message: err.Error()}
	}
	return nil
}

// ToTLSProfile 将领域模型转换为运行时使用的 tlsfingerprint.Profile
// ALPN nil inherits the default; an explicit empty slice disables ALPN.
func (p *TLSFingerprintProfile) ToTLSProfile() *tlsfingerprint.Profile {
	return (&tlsfingerprint.Profile{
		Name:                p.Name,
		EnableGREASE:        p.EnableGREASE,
		ShuffleExtensions:   p.ShuffleExtensions,
		HTTP2:               p.HTTP2,
		CipherSuites:        p.CipherSuites,
		Curves:              p.Curves,
		PointFormats:        p.PointFormats,
		SignatureAlgorithms: p.SignatureAlgorithms,
		ALPNProtocols:       p.ALPNProtocols,
		SupportedVersions:   p.SupportedVersions,
		KeyShareGroups:      p.KeyShareGroups,
		PSKModes:            p.PSKModes,
		Extensions:          p.Extensions,
	}).Clone()
}
